### ISSUE TITLE
"rules process last" should by default assign the conditional subs to both rclt & calt

### ISSUE THREAD

**Issue by twardoch** [fonttools/fonttools#3004]
Do we know why DesignSpace "rules process last" assigns the conditional glyph substitutions only to one feature (rclt) and you can use a lib key to choose another feature like calt? 

It could easily produce a Feature variations lookup and then assign it to several features (like rclt & calt), to maximize engine compatibility. The lookup will be processed once anyway if it's just one lookup.

**Comment by behdad** [comment]
What's the point of adding them to `rclt` if you're also adding to `calt`?

**Comment by behdad** [comment]
Any engine that supports `rclt` also supports `calt`. I suppose the only benefit of adding to both is that on engines that support `rclt`, you wouldn't be able to turn these off by `calt`. So yeah it might be considered an improvement.

**Comment by anthrotype** [comment]
> What's the point of adding them to rclt if you're also adding to calt?

I presume some buggy software refuses to apply rclt but works if feature is named calt.. not making names here.

**Comment by anthrotype** [comment]
yeah, there's a duplicate issue here which explains the rationale of registering to both:

https://github.com/fonttools/fonttools/issues/2050

**Comment by anthrotype** [comment]
this was the conclusion last time we discussed this

> So (if I'm getting this straight), the desire to do both calt and rclt simultaneously is to work around an Adobe bug. I don't see why the designspace format should even be accomodating this. This is a typical case for a gftools-like afterburner tool.

I propose to close unless Adam puts forward more convincing arguments?

**Comment by twardoch** [comment]
Please point me to the resource that indicates that `rclt` should be on by default for all scripts. 

So far I see the [`rclt` description](https://learn.microsoft.com/en-us/typography/opentype/spec/features_pt#rclt) saying: 

> Application interface: In recommended usage, this feature triggers substitutions required for correct display of certain scripts. It should be applied in the appropriate contexts, as determined by script-specific processing requirements. When processing lookups, context before or after the glyph sequence to which the feature is applied must be considered.

The [MS spec for Arabic Script](https://learn.microsoft.com/en-us/typography/script-development/arabic) does list `rclt` among the on-by default features but the [MS spec for Standard Scripts](https://learn.microsoft.com/en-us/typography/script-development/standard) does not. 

If Adobe does not enable `rclt` for all scripts — what’s the basis for claiming that Adobe’s implementation is »buggy«?

**Comment by behdad** [comment]
IIRC in 2019ish OpenType meeting it was decided to apply `rclt` by default. It seems like it never made it to the spec.

Also, the fact that:

> UI suggestion: Control of this feature should not generally be exposed to the user.

> Script/language sensitivity: May be used for any script, but is especially important for many styles of Arabic.

suggest that this should be on by default.

cc @pgconstable

**Comment by twardoch** [comment]
It's good that something was decided in a meeting, but how would shaper implementers who didn't attend the meeting know this? 😃

**Comment by twardoch** [comment]
What I meant is: this whole notion of "on by default" is just very muddy. Script-specific, whatever-specific. In fact, we could just have one `reqd` feature that would do ALL required processing for GSUB & GPOS — and only the shaper-context features like `init` would need to be separate. Why `rvrn` got added is beyond me (it could be part of `ccmp`). Why we have `rclt` and `rlig` separately, is also beyond me. Or `kern` & `dist`. 

But since it's all a mess, and the reality is that de facto we’ve so been going by »implementation-driven spec« (first Uniscribe, then HarfBuzz), I don't see why the Adobe (or any other) implementation is commented as »buggy« while MS (and HB) are somehow »gold«. 😎

**Comment by behdad** [comment]
> Why `rvrn` got added is beyond me

Because MS wanted it. Their argument was that MS Office and other apps process the font to decide whether the font does any "complex" processing. If it doesn't, they send it through a fastpass. They wanted to be able to still do that even if the font has feature-variations... ¯\_(ツ)_/¯

**Comment by khaledhosny** [comment]
> »implementation-driven spec«

That is exactly what it is. Since there is no spec, whatever is the dominant implementation that is the de facto standard and everyone has to copy it. If your implementation is different than what everyone else’s, you are the buggy one. The fact that Adobe has neglected their OpenType implementation(s) for so long that it no longer does what everyone else is doing, that is on them.

We can do the standards-compliance arguing is there is a standard, but OpenType does not have a written standard, and never did, it has de facto standard implementation and that is a known fact.

**Comment by twardoch** [comment]
There are large user groups for whom Adobe is the leading implementation. There is a big divide: we have the i18n-focused no-UI implementations which do well in complex scripts but poorly on exposing discretionary features, and we have implementers that do well in exposing access to user-controlled features but lag on the auto stuff. Like it or not, Adobe is the 2nd, and it has authored half of the feature descriptions. 

Yes, the Adobe situation is annoying. But without Adobe, we wouldn't have swashes, oldstyle features or stylistic sets in OT, because the MS folks didn't care for those at all.

**Comment by twardoch** [comment]
What I mean to say is: Adobe is not just sone random marginal 3rd party to OT. They shipped the first desktop app to expose user-controlled features (InDesign), did a lot to promote OpenType as a brand/concept, they authored tons of features and produced Adobe Originals fonts that excel at implementing then, commissioned high-quality nonLatin designs like Adobe Arabic, were first to ship desktop apps to have sliders for VF support, put resources into OTSVG (spec & implementation). Yes, they made mistakes, but overall they put their money where their mouth was. Not mentioning Adobe FDK & Free type. They didn't do "perfect", but they also didn't produce hot air & no substance. 

Somehow I don't feel comfortable in saying "oh it's a Adobe bug we don't care" while using the Adobe feature definition syntax to author font sources at the same time.

**Comment by twardoch** [comment]
I forgot: they paid EvB to opensource MutatorMath & to make .designdpace wdth we're happily using today.

**Comment by twardoch** [comment]
Please don't go into the "I have privileged access to the Chromium devs do I can make things shipped the way I want, so I don't care about the others" place, is not healthy :)

**Comment by twardoch** [comment]
What I really mean to say 😀 is that I'm not a friend of making hacks that go decidedly against the spirit of a concept. But ever since the old days, since Adobe Illustrator had UI for salt but not ssXX and Adobe InDesign had UI for ssXX but not salt, I know that engine implementers will always do some subset of the spec. As in: Chromium not implementing OT+SVG even if it's in the spec. 

Whenever I advised people on features in fonts, I always said: make your GSUB LookupType 1 subs into ssXX, and then make the equivalent GSUB LookupType 3 subs in salt. Is it wrong? I don't think so. 

There is no COMPLETE implementation of OT.

We always have a PARTIAL implementation. So what should we do? IMO fonts should try to service as many partial implementations as long as they don't BREAK some implementation A in order to service implementation B.

Does assigning the same feature variation lookups to both rclt & calt break any implementation? No. Will it service a larger group of end-users if these feature variations will go into both, in a good way? Yes. Is the overhead if doing both large? No.

So let's do it. That's my argument :)

**Comment by anthrotype** [comment]
> produce a Feature variations lookup and then assign it to several features (like rclt & calt)

this is now possible after #3360, the featureVarsFeatureTag designspace lib key now takes a comma-separated list of tags:

https://github.com/fonttools/fonttools/blob/0b05cec0a345fc9d912170f1fa811215ae54579c/Tests/varLib/data/FeatureVarsCustomTag.designspace#L71-L76
